File: sdks/python-sdk/quantaureum/utils/address.py
# Quantaureum Python SDK source, version 1.0.0.
"""Address utilities for the SDK.

Property 7: Address Validation Correctness
For any string, is_address should return True only for valid 40-character hex strings
with 0x prefix (42 chars total).
Validates: Requirements 5.3
"""

import re
from typing import Any, Union

from eth_hash.auto import keccak as eth_keccak

from ..exceptions import ValidationError


# Regex pattern for valid Ethereum-style addresses
ADDRESS_PATTERN = re.compile(r"^0x[0-9a-fA-F]{40}$")


def is_address(address: Any) -> bool:
    """Check if a value is a valid Ethereum-style address.

    A valid address is a 42-character string starting with '0x' followed by
    40 hexadecimal characters.

    Args:
        address: The value to check.

    Returns:
        True if the value is a valid address, False otherwise.

    Examples:
        >>> is_address('0x742d35Cc6634C0532925a3b844Bc9e7595f1b2b1')
        True
        >>> is_address('0x123')
        False
        >>> is_address(123)
        False
    """
    if not isinstance(address, str):
        return False

    return bool(ADDRESS_PATTERN.match(address))


def get_address(address: str) -> str:
    """Get the checksummed version of an address (EIP-55).

    Args:
        address: The address to checksum.

    Returns:
        The checksummed address.

    Raises:
        ValidationError: If the address is invalid.

    Examples:
        >>> get_address('0x742d35cc6634c0532925a3b844bc9e7595f1b2b1')
        '0x742d35Cc6634C0532925a3b844Bc9e7595f1b2b1'
    """
    if not isinstance(address, str):
        raise ValidationError(
            message=f"Expected string, got {type(address).__name__}",
            field="address",
            value=address,
            expected="string",
        )

    # Normalize to lowercase and check format
    lower_address = address.lower()
    if not ADDRESS_PATTERN.match(lower_address):
        raise ValidationError(
            message=f"Invalid address format: {address}",
            field="address",
            value=address,
            expected="0x followed by 40 hex characters",
        )

    # Remove 0x prefix for hashing
    address_bytes = lower_address[2:].encode("ascii")
    hash_bytes = eth_keccak(address_bytes)
    hash_hex = hash_bytes.hex()

    # Apply EIP-55 checksum
    checksummed = "0x"
    for i, char in enumerate(lower_address[2:]):
        if char in "0123456789":
            checksummed += char
        else:
            # If the corresponding nibble in the hash is >= 8, uppercase
            if int(hash_hex[i], 16) >= 8:
                checksummed += char.upper()
            else:
                checksummed += char

    return checksummed


def compute_address(public_key: Union[str, bytes]) -> str:
    """Compute an address from a public key.

    The address is the last 20 bytes of the Keccak-256 hash of the public key.

    Args:
        public_key: The public key as hex string (with or without 0x prefix)
                   or bytes. Should be 64 bytes (uncompressed without prefix)
                   or 65 bytes (with 0x04 prefix).

    Returns:
        The checksummed address.

    Raises:
        ValidationError: If the public key is invalid.

    Examples:
        >>> compute_address('0x04...')  # 65 bytes with prefix
        '0x...'
    """
    if isinstance(public_key, str):
        # Remove 0x prefix if present
        if public_key.startswith(("0x", "0X")):
            public_key = public_key[2:]
        try:
            public_key = bytes.fromhex(public_key)
        except ValueError as e:
            raise ValidationError(
                message=f"Invalid public key hex: {e}",
                field="public_key",
                value=public_key,
                expected="valid hex string",
            ) from e

    if not isinstance(public_key, bytes):
        raise ValidationError(
            message=f"Expected str or bytes, got {type(public_key).__name__}",
            field="public_key",
            value=public_key,
            expected="str or bytes",
        )

    # Handle different public key formats
    if len(public_key) == 65:
        # Uncompressed public key with 0x04 prefix
        if public_key[0] != 0x04:
            raise ValidationError(
                message="65-byte public key must start with 0x04",
                field="public_key",
                value=public_key,
                expected="0x04 prefix for uncompressed key",
            )
        public_key = public_key[1:]  # Remove prefix
    elif len(public_key) != 64:
        raise ValidationError(
            message=f"Invalid public key length: {len(public_key)}",
            field="public_key",
            value=public_key,
            expected="64 or 65 bytes",
        )

    # Hash the public key and take last 20 bytes
    hash_bytes = eth_keccak(public_key)
    address_bytes = hash_bytes[-20:]

    # Return checksummed address
    raw_address = "0x" + address_bytes.hex()
    return get_address(raw_address)


def is_checksum_address(address: str) -> bool:
    """Check if an address has a valid EIP-55 checksum.

    Args:
        address: The address to check.

    Returns:
        True if the address has a valid checksum, False otherwise.

    Examples:
        >>> is_checksum_address('0x742d35Cc6634C0532925a3b844Bc9e7595f1b2b1')
        True
        >>> is_checksum_address('0x742d35cc6634c0532925a3b844bc9e7595f1b2b1')
        False
    """
    if not is_address(address):
        return False

    try:
        return get_address(address) == address
    except ValidationError:
        return False


def zero_address() -> str:
    """Return the zero address.

    Returns:
        The zero address (0x0000...0000).
    """
    return "0x" + "0" * 40


__all__ = [
    "is_address",
    "get_address",
    "compute_address",
    "is_checksum_address",
    "zero_address",
]

File: sdks/ts-sdk/src/utils/address.ts
// Quantaureum TypeScript SDK source, version 1.0.0.
/**
 * Address utilities for Quantaureum SDK
 * @module utils/address
 */

import { keccak_256, sha3_256 } from '@noble/hashes/sha3';
import { ValidationError, ErrorCode } from '../errors';
import { fromHex, hexlify, isHexString } from './hex';

/**
 * Dilithium3 public key size — 1952 bytes (project cryptographic contract).
 * Used to validate computeAddress inputs. Kept as a local constant instead of
 * importing from the quantum-crypto module to keep utils/address decoupled
 * from the WASM loader (which is a much heavier dependency).
 */
const DILITHIUM3_PUBLIC_KEY_SIZE = 1952;

/**
 * Regular expression for validating Ethereum-style addresses
 * 40 hex characters with 0x prefix (42 chars total)
 */
const ADDRESS_REGEX = /^0x[0-9a-fA-F]{40}$/;

/**
 * Check if a string is a valid Ethereum-style address
 * @param address - String to check
 * @returns True if valid address format
 */
export function isAddress(address: unknown): address is string {
  if (typeof address !== 'string') {
    return false;
  }
  return ADDRESS_REGEX.test(address);
}

/**
 * Get the checksum address (EIP-55)
 * @param address - Address to checksum
 * @returns Checksummed address
 * @throws ValidationError if address is invalid
 */
export function getAddress(address: string): string {
  if (typeof address !== 'string') {
    throw new ValidationError(
      'Address must be a string',
      'address',
      address,
      ErrorCode.INVALID_ADDRESS,
      '0x-prefixed 40 character hex string'
    );
  }

  // Normalize to lowercase without 0x prefix
  let addr = address.toLowerCase();
  if (addr.startsWith('0x')) {
    addr = addr.slice(2);
  }

  // Validate length and characters
  if (addr.length !== 40 || !/^[0-9a-f]{40}$/.test(addr)) {
    throw new ValidationError(
      `Invalid address: ${address}`,
      'address',
      address,
      ErrorCode.INVALID_ADDRESS,
      '0x-prefixed 40 character hex string'
    );
  }

  // Compute checksum using keccak256 of lowercase address
  const hash = keccak_256(new TextEncoder().encode(addr));
  const hashHex = hexlify(hash).slice(2); // Remove 0x prefix

  // Apply checksum
  let checksumAddress = '0x';
  for (let i = 0; i < 40; i++) {
    // If the corresponding hash nibble is >= 8, uppercase the character
    const hashChar = hashHex[i];
    const addrChar = addr[i];
    if (hashChar !== undefined && addrChar !== undefined) {
      const hashNibble = parseInt(hashChar, 16);
      if (hashNibble >= 8) {
        checksumAddress += addrChar.toUpperCase();
      } else {
        checksumAddress += addrChar;
      }
    }
  }

  return checksumAddress;
}

/**
 * Compute address from a Dilithium3 public key (1952 bytes).
 *
 * R40-P0-02 (2026-08-03): the SDK now derives addresses the same way the
 * Quantaureum node and the Quantaureum extension wallet do —
 *   `address = '0x' + hexlify(NIST sha3_256(publicKey)[12:32])`
 * Previously this function hashed the secp256k1 public key with
 * Keccak-256 (Ethereum) and took the last 20 bytes, which produced addresses
 * that did NOT match any account on the Quantaureum chain. The fix switches
 * the hash family to NIST SHA3-256 and accepts the Dilithium3 public key
 * length (1952 bytes). The 65-byte / 64-byte ECDSA pubkey paths are removed
 * (no longer relevant after the Dilithium3 migration); EIP-55 checksum
 * encoding is preserved via getAddress() for backwards compatibility with
 * the address-string representation used throughout the SDK.
 *
 * @param publicKey - Dilithium3 public key (1952 bytes), as `Uint8Array` or
 *                    `0x`-prefixed hex string (3904 hex chars).
 * @returns The EIP-55 checksummed Quantaureum address
 * @throws ValidationError if the public key length is not 1952 bytes
 */
export function computeAddress(publicKey: string | Uint8Array): string {
  let pubKeyBytes: Uint8Array;

  if (typeof publicKey === 'string') {
    if (!isHexString(publicKey)) {
      throw new ValidationError(
        'Public key must be a valid hex string',
        'publicKey',
        publicKey,
        ErrorCode.INVALID_ARGUMENT,
        '1952-byte (3904-hex) Dilithium3 public key',
      );
    }
    pubKeyBytes = fromHex(publicKey);
  } else if (publicKey instanceof Uint8Array) {
    pubKeyBytes = publicKey;
  } else {
    throw new ValidationError(
      'Public key must be a string or Uint8Array',
      'publicKey',
      publicKey,
      ErrorCode.INVALID_ARGUMENT,
      'string or Uint8Array',
    );
  }

  if (pubKeyBytes.length !== DILITHIUM3_PUBLIC_KEY_SIZE) {
    throw new ValidationError(
      `Invalid public key length: ${pubKeyBytes.length} bytes (expected ${DILITHIUM3_PUBLIC_KEY_SIZE} for Dilithium3)`,
      'publicKey',
      '[REDACTED]',
      ErrorCode.INVALID_PUBLIC_KEY,
      `exactly ${DILITHIUM3_PUBLIC_KEY_SIZE} bytes`,
    );
  }

  // NIST SHA3-256 of the raw Dilithium3 public key, take bytes [12:32] (the
  // last 20 bytes of the 32-byte digest). Match `types.AddressFromPublicKeyE`
  // and the wallet's `Dilithium3.deriveAddress` byte-for-byte.
  const hash = sha3_256(pubKeyBytes);
  const addressBytes = hash.slice(12, 32);
  const address = hexlify(addressBytes);

  // Return checksummed address (EIP-55 using keccak_256 of the lowercase
  // address string — checksum is orthogonal to address derivation).
  return getAddress(address);
}

/**
 * Check if an address has a valid checksum (EIP-55)
 * @param address - Address to check
 * @returns True if checksum is valid or address is all lowercase/uppercase
 */
export function isChecksumAddress(address: string): boolean {
  if (!isAddress(address)) {
    return false;
  }

  // All lowercase or all uppercase is valid (no checksum)
  const addr = address.slice(2);
  if (addr === addr.toLowerCase() || addr === addr.toUpperCase()) {
    return true;
  }

  // Check if checksum matches
  try {
    return getAddress(address) === address;
  } catch {
    return false;
  }
}

/**
 * Compare two addresses for equality (case-insensitive)
 * @param a - First address
 * @param b - Second address
 * @returns True if addresses are equal
 */
export function addressEquals(a: string, b: string): boolean {
  if (!isAddress(a) || !isAddress(b)) {
    return false;
  }
  return a.toLowerCase() === b.toLowerCase();
}

/**
 * Zero address constant
 */
export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';

/**
 * Check if an address is the zero address
 * @param address - Address to check
 * @returns True if address is the zero address
 */
export function isZeroAddress(address: string): boolean {
  return addressEquals(address, ZERO_ADDRESS);
}
